src/cluster/agglomerative.rs

```diff
@@ -0,0 +1,219 @@
+//! #  Agglomerative hierarchical clustering
+//! 
+//! ## Definition
+//! SAHN (sequential, agglomerative, hierarchic, nonoverlapping methods)
+//!
+//! as defined in https://arxiv.org/pdf/1109.2378.pdf **by Müllner, 2011**
+//! 
+//! > Agglomerative clustering schemes start from the partition of
+//! > the data set into singleton nodes and merge step by step the current pair of mutually closest
+//! > nodes into a new node until there is one final node left, which comprises the entire data set.
+//! > Various clustering schemes share this procedure as a common definition, but differ in the way
+//! > in which the measure of inter-cluster dissimilarity is updated after each step. The seven most
+//! > common methods are termed single, complete, average (UPGMA), weighted (WPGMA, McQuitty), Ward,
+//! > centroid (UPGMC) and median (WPGMC) linkage.
+//! 
+//! ## Algorithms:
+//! > The specific class of clustering algorithms which is dealt with in this paper has been
+//! > characterized by the acronym SAHN (sequential, agglomerative, hierarchic, nonoverlapping methods)"
+//! 
+//! * (availalbe in ver. 0.2) `PRIMITIVE_CLUSTERING`, as in Fig.1 in Müllner, 2011
+//! * `GENERIC_LINKAGE` Anderberg, 1973 and later improvements
+//! * `NN_CHAIN_LINKAGE` Murtagh, 1985 and later improvements
+//! * (to be implemented) MST-linkage` (The single linkage algorithm, aka fastcluster), as in Fig.6 in Müllner, 2011
+//! 
+//! `MST-linkage` is an implmenetation of `MST-linkage-core` plus two post-processing steps:
+//! * Sort by distance
+//! * LABEL (aka union-find), as in Fig.5 in Müllner, 2011
+//! 
+//! 
+//! Example:
+//!
+//! ```
+//! use smartcore::linalg::naive::dense_matrix::*;
+//! use smartcore::cluster::agglomerative::*;
+//!
+//! // Iris data
+//! let x = DenseMatrix::from_2d_array(&[
+//!            &[5.1, 3.5, 1.4, 0.2],
+//!            &[4.9, 3.0, 1.4, 0.2],
+//!            &[4.7, 3.2, 1.3, 0.2],
+//!            &[4.6, 3.1, 1.5, 0.2],
+//!            &[5.0, 3.6, 1.4, 0.2],
+//!            &[5.4, 3.9, 1.7, 0.4],
+//!            &[4.6, 3.4, 1.4, 0.3],
+//!            &[5.0, 3.4, 1.5, 0.2],
+//!            &[4.4, 2.9, 1.4, 0.2],
+//!            &[4.9, 3.1, 1.5, 0.1],
+//!            &[7.0, 3.2, 4.7, 1.4],
+//!            &[6.4, 3.2, 4.5, 1.5],
+//!            &[6.9, 3.1, 4.9, 1.5],
+//!            &[5.5, 2.3, 4.0, 1.3],
+//!            &[6.5, 2.8, 4.6, 1.5],
+//!            &[5.7, 2.8, 4.5, 1.3],
+//!            &[6.3, 3.3, 4.7, 1.6],
+//!            &[4.9, 2.4, 3.3, 1.0],
+//!            &[6.6, 2.9, 4.6, 1.3],
+//!            &[5.2, 2.7, 3.9, 1.4],
+//!            ]);
+//!
+//! let primitive_clustering = PrimitiveClustering::fit(&x, 2, Default::default()).unwrap(); // Fit to data, 2 clusters
+//!
+use std::cmp;
+use crate::math::num::RealNumber;
+use smartcore::math::distance::euclidian::Euclidian;
+
+
+pub trait SAHNClustering<T: RealNumber> {
+  pub fn fit<M: Matrix<T>>(
+    data: &M,
+    k: usize,
+  ) -> Result<ClusterResult<T>, Failed>;
```

Comment:
We should return Self here to be inline with the rest of the algorithms in SmartCore

Reply:
Unless this class is not public-facing and you plan to add another class that will be similar in design to KMeans